### PR TITLE
Feature Completeness for Compile-to-C++

### PR DESCRIPTION
**Pull request by vpax**
This PR updates `-O gen-C++` functionality to have feature completeness with respect to (non-deprecated) interpreted execution.  The main missing piece has been support for `when` statements, though it also includes some separate bug fixes, and some helper scripts to facilitate future maintenance by running on the test suite.

The one (slightly) subtle part is that `Trigger` objects previously used to associate delayed values with `const CallExpr*`'s.  Now these associations are with `const void*`'s - because all that actually matters is to have a unique opaque value, and for compiled-to-C++ code there are no `CallExpr*`'s any more.  This manifests in `Trigger` now having a notion of an "association" value, which defaults to any `CallExpr*` bound to the `Trigger`, but also can be set directly.  I've marked the previous `Trigger` API as deprecated.  The migration is simple - either cast to `const void*`, or switch from using `Trigger::GetCall()` to instead use `Trigger::GetTriggerAssoc()`.

### CODE REVIEW

File: src/script_opt/CPP/RuntimeOps.h [hunk 1]
```diff
@@ -41,6 +49,20 @@ inline ValPtr invoke__CPP(Func* f, std::vector<ValPtr> args, Frame* frame)
...
+// last argument is the address of the calling function; we just need
+// it to be distinct to the call, so we can associate a Trigger cache
+// with it.
+extern ValPtr when_invoke__CPP(Func* f, std::vector<ValPtr> args, Frame* frame, void* caller_addr);
+
+// Thrown when a call inside a "when" delays.
+class DelayedCallException : public InterpreterException
+	{
+public:
+	DelayedCallException() { }
```
Comment (timwoj) [inline src/script_opt/CPP/RuntimeOps.h:63]: No need to do this. It'll get automatically generated. If you're set on it being there, change `{}` to `= default`.

File: src/Expr.h [hunk 2]
```diff
@@ -1059,6 +1061,13 @@ extern VectorValPtr vector_bool_select(VectorTypePtr vt, const VectorVal* v1, co
 // indices to select).
 extern VectorValPtr vector_int_select(VectorTypePtr vt, const VectorVal* v1, const VectorVal* v2);
 
+// The following is used for index expressions that occur inside "when"
+// clauses.  It tracks all the results produced by evaluating indexing
+// aggregates, so that if any of them are Modifiable(), the associated
+// Trigger can register interest in changes to them.
+//
+// One Fine Day we should do the equivalent for accessing fields in records,
```
[the remark this answers is not in the document]
Comment (vpax) [reply src/Expr.h:1069]: Good thought, will do

File: src/script_opt/CPP/maint/README [hunk 3]
```diff
@@ -3,6 +3,10 @@ This is a collection of scripts to support maintenance of -O gen-C++
 suite, and are currently tailored for the lead maintainer's own environment.
 The scripts all assume you're running them from build/ .
 
+If you make changes to the scripts, format them using
+
+	shfmt -w -i 4 -ci *.sh
```
[the remark this answers is not in the document]
Comment (vpax) [reply src/script_opt/CPP/maint/README:8]: Added a note to that effect

File: src/Trigger.h [hunk 4]
```diff
@@ -86,7 +87,18 @@ class Trigger final : public Obj, public notifier::detail::Receiver
 	// void*'s so that the value can be associated with either a CallExpr
 	// (for interpreted execution) or a C++ function (for compiled-to-C++).
 	bool Cache(const void* obj, Val* val);
-	Val* Lookup(const void*);
+	Val* Lookup(const void* obj);
```
[the remark this answers is not in the document]
Comment (vpax) [reply src/Trigger.h:90]: Yeah.  The basic problem is that the key can be either a pointer-to-a-`CallExpr` AST node (so not even a C++ function!) or a pointer-to-a-C++-function, and there's no commonality between those two other than the fact that they're both pointers.  At least with `const`, the new approach makes it clear that no one is going to do anything with the pointer other than stare at its bit pattern.

File: src/script_opt/CPP/RuntimeOps.h (hunk 1 above)
Comment (vpax) [reply src/script_opt/CPP/RuntimeOps.h]: Got it, thx.  All of the above now addressed with the latest pushes (which also include unrelated NEWS items, for convenience).

**Review by timwoj — APPROVED** [review]

**Pull request merged**